src/gui/folder.cpp

```diff
@@ -1218,6 +1223,74 @@ void Folder::slotHydrationDone()
     emit syncStateChange();
 }
 
+void Folder::slotNewShellExtensionConnection()
+{
+    const auto newConnection = _shellExtensionsServer.nextPendingConnection();
+
+    connect(newConnection, &QLocalSocket::errorOccurred, this, [newConnection](QLocalSocket::LocalSocketError socketError) {
+        qCCritical(lcFolder) << "Shell extension socket error: " << socketError << " : " << newConnection->errorString();
+    });
+
+    const auto disconnectAndCloseSocket = [newConnection, this]() {
+        connect(newConnection, &QLocalSocket::disconnected, this, [newConnection] {
+            newConnection->close();
+            newConnection->deleteLater();
+        });
+        newConnection->disconnectFromServer();
+    };
+
+    const auto sendEmptyData = [newConnection]() {
+        newConnection->write({});
+        newConnection->waitForBytesWritten();
+    };
+
+    if (!newConnection->waitForReadyRead()) {
+        disconnectAndCloseSocket();
+        return;
+    }
+
+    const auto receivedMessage = QJsonDocument::fromJson(newConnection->readAll()).toVariant().toMap();
+
+    const auto thumbnailRequestMessage = receivedMessage.value(CfApiShellExtensions::Protocol::ThumbnailProviderRequestKey).toMap();
+
+    const auto thumbnailFilePath = thumbnailRequestMessage.value(CfApiShellExtensions::Protocol::ThumbnailProviderRequestFilePathKey).toString();
+    const auto thumbnailFileSize = thumbnailRequestMessage.value(CfApiShellExtensions::Protocol::ThumbnailProviderRequestFileSizeKey).toMap();
+
+    if (thumbnailFilePath.isEmpty() || thumbnailFileSize.isEmpty()) {
+        disconnectAndCloseSocket();
+        return;
+    }
+
+    const auto fileInfo = QFileInfo(thumbnailFilePath);
+    const auto filePathRelative = QFileInfo(thumbnailFilePath).canonicalFilePath().remove(path());
+
+    SyncJournalFileRecord record;
+    if (!_journal.getFileRecord(filePathRelative, &record)) {
+        sendEmptyData();
+        disconnectAndCloseSocket();
+        return;
+    }
+
+    QUrlQuery queryItems;
+    queryItems.addQueryItem("fileId", record._fileId);
+    queryItems.addQueryItem("x", thumbnailFileSize.value("x").toString());
+    queryItems.addQueryItem("y", thumbnailFileSize.value("y").toString());
+    const QUrl jobUrl = Utility::concatUrlPath(accountState()->account()->url(), "core/preview", queryItems);
```

Comment:
this URL may report a 404 and is missing an additional `index.php` element in the path before `core/preview`